protoc-gen-gogo/generator/helper.go

```diff
@@ -220,6 +220,55 @@ func IsMap(file *descriptor.FileDescriptorProto, field *descriptor.FieldDescript
 	return msg.GetOptions().GetMapEntry()
 }
 
+func (g *Generator) GetMapKeyField(field, keyField *descriptor.FieldDescriptorProto) *descriptor.FieldDescriptorProto {
+	if !gogoproto.IsCastKey(field) {
+		return keyField
+	}
+	keyField = cloneFieldExtensions(keyField)
+	keyType := gogoproto.GetCastKey(field)
+	if err := proto.SetExtension(keyField.Options, gogoproto.E_Casttype, &keyType); err != nil {
+		g.Fail(err.Error())
+	}
+	return keyField
+}
+
+func (g *Generator) GetMapValueField(field, valField *descriptor.FieldDescriptorProto) *descriptor.FieldDescriptorProto {
+	if !gogoproto.IsCastValue(field) && gogoproto.IsNullable(field) {
+		return valField
+	}
+	valField = cloneFieldExtensions(valField)
+	if valType := gogoproto.GetCastValue(field); len(valType) > 0 {
+		if err := proto.SetExtension(valField.Options, gogoproto.E_Casttype, &valType); err != nil {
+			g.Fail(err.Error())
+		}
+	}
+
+	nullable := gogoproto.IsNullable(field)
+	if err := proto.SetExtension(valField.Options, gogoproto.E_Nullable, &nullable); err != nil {
+		g.Fail(err.Error())
+	}
+	return valField
+}
+
+// cloneFieldExtensions creates a copy of field where field.Options.XXX_extensions is safe to mutate.
+func cloneFieldExtensions(field *descriptor.FieldDescriptorProto) *descriptor.FieldDescriptorProto {
```

Comment:
Have you tried the proto.Clone function?
It should work for the FieldDescriptor, since it has not been modified by nullable=false and customtypes etc.